packages/flutter/lib/src/widget_preview/widget_preview.dart

```diff
@@ -0,0 +1,302 @@
+  const _InteractiveViewerWrapper({
+    required this.child,
+    required this.transformationController,
+  });
+
+  final Widget child;
+  final TransformationController transformationController;
+
+  @override
+  Widget build(BuildContext context) {
+    return InteractiveViewer(
+      transformationController: transformationController,
+      scaleEnabled: false,
+      child: child,
+    );
+  }
+}
+
+/// Wrapper applying a custom render object to force constraints on
+/// unconstrained widgets.
+@visibleForTesting
+class WidgetPreviewWrapper extends SingleChildRenderObjectWidget {
+  @visibleForTesting
+  // ignore: public_member_api_docs
+  const WidgetPreviewWrapper({
+    super.key,
+    super.child,
+    required this.previewerConstraints,
+  });
+
+  /// The ratio of the max height provided by a parent
+  /// [WidgetPreviewerWindowConstraints] that an unconstrained child should
+  /// be allowed to occupy.
+  @visibleForTesting
+  static const double unconstrainedChildScalingRatio = 0.5;
+
+  /// The size of the previewer render surface.
+  final BoxConstraints previewerConstraints;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return WidgetPreviewWrapperBox(
+      previewerConstraints: previewerConstraints,
+      child: null,
+    );
+  }
+
+  @override
+  void updateRenderObject(
+    BuildContext context,
+    WidgetPreviewWrapperBox renderObject,
+  ) {
+    renderObject.setPreviewerConstraints(previewerConstraints);
+  }
+}
+
+/// Custom render box that forces constraints onto unconstrained widgets.
+@visibleForTesting
+class WidgetPreviewWrapperBox extends RenderShiftedBox {
+  // ignore: public_member_api_docs
+  WidgetPreviewWrapperBox({
+    required RenderBox? child,
+    required BoxConstraints previewerConstraints,
+  })  : _previewerConstraints = previewerConstraints,
+        super(child);
+
+  BoxConstraints _constraintOverride = const BoxConstraints();
+  BoxConstraints _previewerConstraints;
+
+  /// Updates the constraints for the child based on changes to the constraints
+  /// provided by the parent [WidgetPreviewerWindowConstraints].
+  void setPreviewerConstraints(BoxConstraints previewerConstraints) {
+    if (_previewerConstraints == previewerConstraints) {
+      return;
+    }
+    _previewerConstraints = previewerConstraints;
+    markNeedsLayout();
+  }
+
+  @override
+  void layout(
+    Constraints constraints, {
+    bool parentUsesSize = false,
+  }) {
+    if (child != null && constraints is BoxConstraints) {
+      double minInstrinsicHeight;
+      try {
+        minInstrinsicHeight = child!.getMinIntrinsicHeight(
+          constraints.maxWidth,
+        );
+      } on Object {
+        minInstrinsicHeight = 0.0;
+      }
+      // Determine if the previewed widget is vertically constrained. If the
+      // widget has a minimum intrinsic height of zero given the widget's max
+      // width, it has an unconstrained height and will cause an overflow in
```

Comment:
I don't think this is true in general. You can write a widget that has a minimum intrinsic height of, say, 25 but if given the chance would like to be as tall as possible.

Also, why if the overflow case only handled for unconstrained height? Wouldn't it also be a problem for unconstrained width?